lib/internal/streams/pipeline.js

```diff
@@ -50,13 +50,30 @@ function destroyer(stream, reading, writing, final, callback) {
       return callback();
     }
 
-    if (!err && reading && !writing && stream.writable) {
-      return callback();
-    }
+    const wState = stream._writableState;
+
+    const writableEnded = stream.writableEnded ||
+      (wState && wState.ended);
+    const writableFinished = stream.writableFinished ||
+      (wState && wState.finished);
+
+    const willFinish = stream.writable ||
+      (writableEnded && !writableFinished);
+    const willEnd = stream.readable;
 
-    if (err || !final || !stream.readable) {
-      destroyImpl.destroyer(stream, err);
+    if (!err) {
+      // First
+      if (reading && !writing && willFinish) {
+        return callback();
+      }
+
+      // Last
+      if (!reading && writing && willEnd) {
+        return callback();
+      }
```

Comment:
Nit: it is possible to reduce the code branches a bit:

```suggestion
    if (!err) {
      // Check if the last stream will end:
      if (!reading) {
        if (writing && stream.readable) {
          return callback();
        }
      // First stream
      } else if (!writing) {
        if (stream.writable) {
          return callback();
        }
        const wState = stream._writableState;
        const writableEnded = stream.writableEnded ||
          (wState && wState.ended);
        const writableFinished = stream.writableFinished ||
          (wState && wState.finished);

        if (writableEnded && !writableFinished)
          return callback();
      }
    }
```

Reply:
@BridgeAR: Any specific reason? This does make it a bit harder to read in my opinon.

Reply:
Code should IMO always only run when necessary. Right now some code is executed that is only needed in some situations.

Reply:
This is not a hot path so readability should IMO come first. I think I can make a compromise 😄. PTAL.